packages/suite/src/actions/suite/metadataActions.ts

```diff
@@ -818,125 +807,487 @@ export const setDeviceMetadataKey =
+    }
+
+    // 5. migration
+    if (!getState().metadata.initiating) {
+        dispatch({ type: METADATA.SET_INITIATING, payload: true });
+    }
+    const migrationResult = await dispatch(handleEncryptionVersionMigration());
+    // failed migration => labeling disabled
+    if (!migrationResult.success) {
+        dispatch({ type: METADATA.SET_INITIATING, payload: false });
+        dispatch({ type: METADATA.SET_EDITING, payload: undefined });
+        dispatch({
+            type: METADATA.SET_FAILED_MIGRATION,
+            payload: {
+                deviceState: device.state!,
+                failed: true,
+            },
+        });
+        dispatch(
+            notificationsActions.addToast({
+                type: 'error',
+                error: `migration failed: ${migrationResult.error}`,
+            }),
+        );
+
+        return;
+    }
+
+    // 6. fetch metadata
+    await dispatch(fetchAndSaveMetadata(device.state));
+
+    // now we may allow user to edit labels. everything is ready, local data is synced with provider
+    if (getState().metadata.initiating) {
+        dispatch({ type: METADATA.SET_INITIATING, payload: false });
+    }
+
+    // 7. if interval for watching provider is not set, create it
+    if (device.state && !fetchIntervals[device.state]) {
+        // todo: possible race condition, user is editing label and at that very moment update arrives. updates to specific entities should be probably discarded in such case?
+        fetchIntervals[device.state] = setInterval(() => {
+            const { device } = getState().suite;
+            if (!getState().suite.online || !device?.state) {
+                return;
+            }
+            dispatch(fetchAndSaveMetadata(device.state));
+        }, METADATA.FETCH_INTERVAL);
+    }
+
+    return true;
+};
+
+export const setEditing = (payload: string | undefined): MetadataAction => ({
+    type: METADATA.SET_EDITING,
+    payload,
+});
+
+export const getLabelableEntitiesDescriptors = () => (dispatch: Dispatch, getState: GetState) => {
+    const { device } = getState().suite;
+
+    if (!device?.state) return [];
+
+    const entitites = dispatch(getLabelableEntities(device.state));
+
+    return entitites
+        .map(entity => {
+            if ('key' in entity) return entity.key;
+            if ('state' in entity && entity.state) return entity.state;
+            throw new Error('entity without unique identifier');
+        })
+        .sort((a, b) => a.localeCompare(b));
+};
+
+export const setEntititesDescriptors = (descriptors: string[]) => (dispatch: Dispatch) => {
+    dispatch({
+        type: METADATA.SET_ENTITIES_DESCRIPTORS,
+        payload: descriptors,
+    });
+};
+
+/**
+ *
+ * @returns files in storage provider split into [[...current][...old]].
+ */
+const getMetadataFiles = () => async (dispatch: Dispatch, getState: GetState) => {
+    const providerInstance = dispatch(
+        getProviderInstance({
+            // todo: huh!
```

Comment:
❓

Reply:
initially my plan was to put correct selector here, but it can be done later. now there may be only zero or one provider in providers array. so just removed the comment.